File: python/mmd_litellm/structured.py
from __future__ import annotations

import json
import re
from typing import Awaitable, Callable, TypeVar

from pydantic import BaseModel, ValidationError

T = TypeVar("T", bound=BaseModel)


def extract_json(text: str) -> str:
    fenced = re.search(r"```json\s*([\s\S]*?)```", text, re.IGNORECASE)
    if fenced is None:
        fenced = re.search(r"```\s*([\s\S]*?)```", text)
    return (fenced.group(1) if fenced else text).strip()


async def call_structured(
    complete: Callable[[str | None], Awaitable[str]],
    schema: type[T],
    *,
    max_repair_attempts: int = 2,
) -> T:
    last_error: str | None = None
    for _attempt in range(max_repair_attempts + 1):
        repair_note = (
            "Your previous output failed JSON schema validation: "
            f"{last_error}. Return corrected JSON only, no prose."
            if last_error
            else None
        )
        text = await complete(repair_note)
        try:
            parsed = json.loads(extract_json(text))
            return schema.model_validate(parsed)
        except (json.JSONDecodeError, ValidationError) as error:
            last_error = str(error)

    raise ValueError(
        "structured output failed schema validation after "
        f"{max_repair_attempts + 1} attempts: {last_error}"
    )

File: packages/model-adapters/src/structured.ts
import type { z } from "zod";

export interface StructuredCallOptions {
  maxRepairAttempts?: number;
}

export function extractJson(text: string): string {
  const fenced =
    text.match(/```json\s*([\s\S]*?)```/i) ?? text.match(/```\s*([\s\S]*?)```/);
  return (fenced ? fenced[1] : text).trim();
}

/**
 * Calls `complete` (a closure that already knows which model/prompt to use),
 * parses the response as JSON, and validates it against `schema`. On parse or
 * validation failure it re-invokes `complete` with a repair note describing
 * what went wrong, up to `maxRepairAttempts` times, instead of failing the
 * whole phase on the first malformed response.
 */
export async function callStructured<T>(
  complete: (repairNote?: string) => Promise<{ text: string }>,
  // Input is widened to `any` (rather than defaulting to T) because schemas
  // with `.default(...)` have an Output type that differs from their Input
  // type; pinning both to T here made TS unify T with the wrong one.
  schema: z.ZodType<T, z.ZodTypeDef, any>,
  opts: StructuredCallOptions = {}
): Promise<T> {
  const maxAttempts = opts.maxRepairAttempts ?? 2;
  let lastError: string | undefined;

  for (let attempt = 0; attempt <= maxAttempts; attempt++) {
    const repairNote = lastError
      ? `Your previous output failed JSON schema validation: ${lastError}. Return corrected JSON only, no prose.`
      : undefined;
    const { text } = await complete(repairNote);

    try {
      const parsed = JSON.parse(extractJson(text));
      const result = schema.safeParse(parsed);
      if (result.success) {
        return result.data;
      }
      lastError = result.error.issues
        .map((i) => `${i.path.join(".") || "(root)"}: ${i.message}`)
        .join("; ");
    } catch (err) {
      lastError = err instanceof Error ? err.message : String(err);
    }
  }

  throw new Error(
    `structured output failed schema validation after ${maxAttempts + 1} attempts: ${lastError}`
  );
}
